File: ncert7/codes/code.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

// Function to calculate PMF using the binomial distribution
void calculate_pmf_binomial(double *pmf) {
    int n = 3; // Number of tosses
    double p = 0.5; // Probability of heads
    for (int k = 0; k <= n; k++) {
        // Calculate binomial coefficient C(n, k) using iterative approach
        double binomial_coefficient = 1;
        for (int i = 1; i <= k; i++) {
            binomial_coefficient *= (double)(n - i + 1) / i;
        }

        // Compute PMF value for k heads
        pmf[k] = binomial_coefficient * pow(p, k) * pow(1 - p, n - k);
    }
}

// Main function
int main() {
    double pmf[4]; // Array to hold PMF values for 0, 1, 2, 3 heads
    calculate_pmf_binomial(pmf);

    // Print the PMF values
    printf("PMF values for 3 coin tosses:\n");
    for (int k = 0; k <= 3; k++) {
        printf("P(%d heads) = %.4f\n", k, pmf[k]);
    }

    return 0;
}

File: ncert7/codes/code.py
import math
import matplotlib.pyplot as plt

# Function to calculate PMF using the binomial distribution
def calculate_pmf_binomial():
    n = 3  # Number of tosses
    p = 0.5  # Probability of heads
    pmf = [0] * (n + 1)  # List to hold PMF values for 0, 1, 2, 3 heads

    for k in range(n + 1):
        # Calculate binomial coefficient C(n, k) using iterative approach
        binomial_coefficient = 1
        for i in range(1, k + 1):
            binomial_coefficient *= (n - i + 1) / i

        # Compute PMF value for k heads
        pmf[k] = binomial_coefficient * (p ** k) * ((1 - p) ** (n - k))
    
    return pmf

# Main execution
if __name__ == "__main__":
    # Calculate the PMF
    pmf_values = calculate_pmf_binomial()

    # Print the PMF values
    print("PMF values for 3 coin tosses:")
    for k in range(4):
        print(f"P({k} heads) = {pmf_values[k]:.4f}")

    # Plotting the PMF
    x = [0, 1, 2, 3]  # Number of heads (0 to 3)
    plt.stem(x, pmf_values, basefmt=" ", use_line_collection=True)
    plt.xlabel('Number of Heads')
    plt.ylabel('Probability')
  
    plt.xticks(x)  # Set x-axis labels to 0, 1, 2, 3
    plt.grid(True)
    plt.show()

